### PR TITLE
refactor: rework `Subeffecting` semantics

### PR DESCRIPTION
**Pull request by magnus-madsen**

### CODE REVIEW

**Comment by magnus-madsen** [comment]
@JonathanStarup Unless something is wrong, I suggest to merge this, and then to iterate.

**Review by mlutze — APPROVED** [review]

**Comment by magnus-madsen** [comment]
@JonathanStarup I asked Matt to approve this so I can iterate, but please do a retroactive review, just to see its OK.

File: main/src/ca/uwaterloo/flix/language/phase/Redundancy.scala [hunk 1]
```diff
@@ -641,7 +642,7 @@ object Redundancy {
           else
             visitExp(exp, env0, rc)
         case CheckedCastType.EffectCast =>
-          if (exp.eff == eff)
+          if (exp.eff == eff && flix.options.xsubeffecting == Disabled)
```
Comment (JonathanStarup) [inline main/src/ca/uwaterloo/flix/language/phase/Redundancy.scala:645]: Why is this here?
